server/src/main/java/org/opensearch/action/bulk/TransportBulkAction.java

```diff
@@ -379,6 +364,80 @@ public void onRejection(Exception rejectedException) {
+        }).filter(Objects::nonNull).toList();
+    }
+
+    /**
+     * Helper method that resolves the correct pipelines for a given list of doc write requests.
+     * Index, update, and delete operations are passed in through the bulk request as doc write requests and
+     * have different interactions with ingest pipelines depending on user parameters.
+     * If pipelines are resolved for a request, the request is mutated with the pipeline information.
+     * @param docWriteRequests the incoming requests from the bulk operation
+     * @param metadata additional metadata used to resolve pipelines, such as index information
+     * @param minNodeVersion the minimum node version of the cluster
+     * @return whether any index request has had a pipeline resolved.
+     */
+    private boolean resolvePipelinesForActionRequests(
+        List<DocWriteRequest<?>> docWriteRequests,
+        Metadata metadata,
+        Version minNodeVersion
+    ) {
+        boolean hasIndexRequestsWithPipelines = false;
+        for (DocWriteRequest<?> actionRequest : docWriteRequests) {
+            // Each index request needs to be evaluated, because this method also modifies the IndexRequest
+            if (actionRequest instanceof UpdateRequest updateRequest) {
+                // Note that when execute pipelines on update existing docs, the full doc is NOT fetched from shard level.
+                // This means the pipeline is only executed on the partial doc, which may not contain all fields.
+                // System ingest pipelines and processors should handle these cases individually.
+                boolean indexRequestHasPipeline = false;
+                IndexRequest upsertRequest = updateRequest.upsertRequest();
+                IndexRequest docRequest = updateRequest.doc();
+                if (upsertRequest != null) {
+                    indexRequestHasPipeline |= ingestService.resolvePipelines(actionRequest, upsertRequest, metadata);
+                }
+                if (docRequest != null) {
+                    if (updateRequest.docAsUpsert()) {
+                        indexRequestHasPipeline |= ingestService.resolvePipelines(actionRequest, docRequest, metadata);
+                    } else {
+                        // In the case when doc as upsert is false or not defined, we only resolve system ingest pipelines.
+                        // This is to preserve backwards compatibility where default and final pipelines are NOT resolved
+                        // and NOT executed on the partial doc passed in from the user request.
+                        // TODO : In the future we want to add configurability to do resolve all pipelines for the doc request
+                        // See https://github.com/opensearch-project/OpenSearch/issues/17742
+                        indexRequestHasPipeline |= ingestService.resolveSystemIngestPipeline(actionRequest, docRequest, metadata);
```

Comment:
Now that we are resolving system ingest pipeline for the request, what are the limitations to resolve it for the ingest pipeline here itself?